Support configuration of entity reporting for V2 quirks

benbancroft opened the pull request:
# Proposed change

This change hooks up the V2 Quirk `report_config` option to the cluster handler created for V2 quirks. It also 'claims' the cluster, which is required for reporting and binding hooks to work upon configuration of the cluster and previously wasn't being done for V2 Quirk clusters.

![Screenshot 2024-10-21 111956](https://github.com/user-attachments/assets/36205b94-3836-4439-bb91-a571c8da050c)

# Additional information

Please refer to https://github.com/zigpy/zigpy/pull/1490 for more information on the motivation behind this patch. This patch will need to be merged after that has been merged and zigpy dependency has been bumped.

benbancroft committed:
Support configuration of reporting for V2 quirks

Note: this requires zigpy export for exposing this via V2 quirk API.

benbancroft committed:
Move ReportingConfig to dataclass

benbancroft committed:
Add test coverage

benbancroft committed:
Change import for ReportingConfig

dmulcahey commented:
needs a rebase

benbancroft commented:
@dmulcahey I've just force pushed a rebase. The tests will continue to fail however, as this patch has dependeny on the zigpy PR, so this will require a further rebase once that merges.

dmulcahey commented:
> @dmulcahey I've just force pushed a rebase. The tests will continue to fail however, as this patch has dependeny on the zigpy PR, so this will require a further rebase once that merges.

My bad.. sorry for the unnecessary ping!!

benbancroft commented:
I've just rebased against zigpy dependency update. All tests should now be passing.

TheJulianJES commented on zha/application/discovery.py:
```diff
@@ -306,6 +308,27 @@ def discover_quirks_v2_entities(self, device: Device) -> None:
+
+                    cluster_handler.__dict__[zha_const.REPORT_CONFIG] = reporting_config
+
+                endpoint.claim_cluster_handlers([cluster_handler])
```
Just noting (for myself) that claiming a cluster handler is what also makes ZHA configure and bind the cluster.
In the future, we may want to have a quirks v2 API for only doing this.

TheJulianJES commented on zha/application/discovery.py:
```diff
@@ -306,6 +308,27 @@ def discover_quirks_v2_entities(self, device: Device) -> None:
+                        ),
+                    )
+
+                    cluster_handler.__dict__[zha_const.REPORT_CONFIG] = reporting_config
```
The attributes are also added to `ZCL_INIT_ATTRS` above this new code block.

If I'm looking at this correctly, we'll have them in both `REPORT_CONFIG` and `ZCL_INIT_ATTRS` and then have duplicates in `uncached`?
https://github.com/zigpy/zha/blob/11cf9ba5d3a47f8733df0c095befac233acb8b73/zha/zigbee/cluster_handlers/__init__.py#L458-L460

We might want to move the new block before the `ZCL_INIT_ATTRS` block and then check in the `ZCL_INIT_ATTRS` block that they're not already in `REPORT_CONFIG` OR just remove them from `ZCL_INIT_ATTRS` in this line?

TheJulianJES replied:
This was solved (in a somewhat different way) with: https://github.com/zigpy/zha/pull/254/commits/69f6242fdd62f0fe1235e8e352d3c35e9a50ef8a

ikruglov commented:
Hi.

Thanks for coming up with a proper solution for the problem of the Ikea VINDSTYRKA sensor's VOC index not being updated.

I want to say that this patch, together with my quirk (https://github.com/ikruglov/zha-device-handlers/commit/c49d8e15120902591100cd8e34ce2778ef525682), works well. I now have a properly updating VOC Index without workarounds, such as manually reading values.

I'd be happy to submit the PR for the quirk, but I thought you may already have one. So, I'll keep it to myself, meanwhile.

Thanks again.

benbancroft committed:
Handle both REPORT_CONFIG and ZCL_INIT_ATTRS

If both ZCL_INIT_ATTRS and REPORT_CONFIG contain
the same attribute, with the current cluster logic,
you end up in a situation where said attribute is
marked as both cacheable and uncacheable at the
same time, leading to it being configured twice.

As they are designed to be mutually exclusive of
each other, let's add the logic to remove attributes
from ZCL_INIT_ATTRS if they are already present in
REPORT_CONFIG.

REPORT_CONFIG will take precedence over ZCL_INIT_ATTRS.

benbancroft commented:
Hi @TheJulianJES 

Many thanks for reviewing this and catching the cacheablity issue - and apologies for my delay in getting back to you on this PR.

I've gone ahead and with the "just remove them from ZCL_INIT_ATTRS in this line" option, as arguably this is really a latent issue that is already possible to expose via old cluster API. I think the intention is for REPORT_CONFIG and ZCL_INIT_ATTRS to be mutually exclusive of each other anyway.

@ikruglov no worries. It certainly would be good to get that quirk in soon as I know it has been around in code review in various forms for a very long time now. Please do go ahead and submit your PR (I wasn't planning to myself).

TheJulianJES commented on zha/zigbee/cluster_handlers/__init__.py:
```diff
@@ -455,9 +455,10 @@ async def async_initialize(self, from_cache: bool) -> None:
-        cached = [a for a, cached in self.ZCL_INIT_ATTRS.items() if cached]
-        uncached = [a for a, cached in self.ZCL_INIT_ATTRS.items() if not cached]
-        uncached.extend([cfg["attr"] for cfg in self.REPORT_CONFIG])
+        cache_config = self.ZCL_INIT_ATTRS.copy()
+        cache_config |= {cfg["attr"]: False for cfg in self.REPORT_CONFIG}
+        cached = [a for a, cached in cache_config.items() if cached]
+        uncached = [a for a, cached in cache_config.items() if not cached]
```
Actually, can we not deduplicate/remove this here, as this might change in a different way in the future and instead just explicitly remove it from `cluster_handler.__dict__[zha_const.ZCL_INIT_ATTRS]` where the `REPORT_CONFIG` is modified in `discovery`?

TheJulianJES commented:
I've opened another PR that also implements parsing the quirks v2 reporting config, add tests, and optimizes how `ZCL_INIT_ATTRS` are set compared to this implementation:
- https://github.com/zigpy/zha/pull/363

I'll draft this PR for now, since we'll likely go with the slightly different implementation I linked.
Still, thanks very much for this PR!

TheJulianJES commented:
Superseded by https://github.com/zigpy/zha/pull/363
